cli/file_watcher.rs

```diff
@@ -13,12 +14,66 @@ use notify::RecursiveMode;
 use notify::Watcher;
 use std::path::PathBuf;
 use std::pin::Pin;
-use tokio::select;
-use tokio::sync::mpsc;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tokio::sync::{mpsc, mpsc::Receiver};
+
+const DEBOUNCE_TIMEOUT_MS: Duration = Duration::from_millis(200);
 
 // TODO(bartlomieju): rename
 type WatchFuture = Pin<Box<dyn Future<Output = Result<(), AnyError>>>>;
 
+struct Debounce {
+  rx: Receiver<Result<NotifyEvent, AnyError>>,
+  debounce_time: Duration,
+  last_event: NotifyEvent,
+}
+
+impl Debounce {
+  fn new(
+    rx: Receiver<Result<NotifyEvent, AnyError>>,
+    debounce_time: Duration,
+  ) -> Self {
+    Self {
+      rx,
+      debounce_time,
+      last_event: Default::default(),
+    }
+  }
+}
+
+impl Stream for Debounce {
+  type Item = NotifyEvent;
+
+  fn poll_next(
+    self: Pin<&mut Self>,
+    _cx: &mut Context,
+  ) -> Poll<Option<Self::Item>> {
+    let mut _self = self.get_mut();
+    let mut timeout = Instant::now();
+    let mut recv = false;
+    loop {
+      if let Ok(result) = _self.rx.try_recv() {
+        if let Ok(event) = result {
+          if event == _self.last_event {
+            // if received event is the same as previous one reset timeout
+            timeout = Instant::now();
+          }
+          // we want to emit only the last received event
+          _self.last_event = event;
+          // event received with success
+          recv = true;
+        }
+      }
+      // if event successfully received and debounce time has passed break and emit last event
+      if recv && timeout.elapsed() >= _self.debounce_time {
+        break;
+      }
+    }
```

Comment:
This is a hot loop; can you add some kind of sleep (10ms?) if no event was received from `rx`?